### PR TITLE
Multiple queue, multiple cluster in one site

### PR DESCRIPTION
**Pull request by sinowood**
Support for multi-queue, multi-cluster configuration in one django site.
API and code changes include:

* Adding `cluster` to async_task() parameters and Task model property
* Adding argument `--name` to qcluster command
* Necessary adjustments to Conf and Broker classes
* Some admin improvements

Examples to run multiple queues (and clusters) in one site:
```
# start qcluster with the default name in QCLUSTER_SETTINGS
python manage.py qcluster

# start qcluster with an alternative cluster name
python manage.py qcluster --name myqueue
```
Task processing rules:
> For async_task() and schedule():
>   if `cluster` is set, only clusters with matching name will run the task or do the schedule;
>   if `cluster` is not set, async_task() will be handled by the default cluster, while schedule() may be handled by any active cluster.

If you do not use multi-queue feature, there should be no difference and requires no changes to your code. only `python manage.py migrate` is needed.

### CODE REVIEW

**Comment by GDay** [comment]
Thanks for the PR. I like it! 

Could you please fix the tests and add some documentation for this? If not, just let me know I can pick it up from here.

**Comment by sinowood** [comment]
OK, I will try to fixed it!

**Comment by msabatier** [comment]
This is an interesting improvement.

One question though
> For async_task() and schedule():
> if `cluster` is set, only clusters with matching name will run the task or do the schedule;
> if `cluster` is not set, async_task() will be handled by the default cluster, while schedule() may be handled by any active cluster.

What is the rational for the difference between tasks and schedules ? Is it by design or due to an implementation constraint ?

I could use this new feature to create priority queues, having one queue to serve short, urgent tasks and another one running long tasks. In this scenario it would be much easier to have schedules run by the default cluster as well (in this case the low priority cluster).

**Comment by sinowood** [comment]
> 

It's due to an implementation constraint. Scheduled task entries reside in ORM, every cluster can query the DB table for its own tasks (as well as tasks ommiting cluster name), and put them in its own queue.
On the other hand, when async_task is execute, a new task is created and directly sent to a task queue, so there must be a destination queue (e.g. redis list_key). In case if cluster name is ommited, the most reasonable destination is the default queue.

It's trival to create priority queues, just do it like this: `async_task(..., cluster='short')`, `schedule(..., cluster='long')`.
When the default broker sees the cluster keyword argument, it will send the task to that corrensponding queue.

**Comment by sinowood** [comment]
I think I have fix the tests, or mostly. If everything is fine I should be updating the docs soon.

**Comment by sinowood** [comment]
Well.. default cluster can be distinguished before configuration, problem can be solved

File: django_q/conf.py [hunk 1]
```diff
@@ -39,10 +39,22 @@ class Conf:
...
-        conf = settings.Q_CLUSTER
+        conf = settings.Q_CLUSTER.copy()
     except AttributeError:
         conf = {}
 
+    _Q_CLUSTER_NAME = os.getenv("Q_CLUSTER_NAME")
+    if _Q_CLUSTER_NAME and _Q_CLUSTER_NAME != conf.get("name") and \
+            _Q_CLUSTER_NAME != conf.get("cluster_name"):
+        conf["cluster_name"] = _Q_CLUSTER_NAME
+        alt_conf = conf.pop("ALT_CLUSTERS")
```
Comment (GDay) [inline django_q/conf.py:50]: This crashes when "ALT_CLUSTERS" has not been configured. Raising an error would probably be better, but it's not a big deal.

**Review by GDay — APPROVED** [review]
@sinowood This is awesome, love it! Works great and code looks good. Thanks a lot! Will merge and release this ASAP.

One small thing that popped up, just as an FYI, but there is no need to get it fixed.

File: django_q/conf.py (hunk 1 above)
Comment (sinowood) [reply django_q/conf.py:50]: I have pushed fixes to this problem in my branch, did I fail to include those fixes in this PR?

**Review by sinowood — COMMENTED** [review]
reviewed

File: django_q/conf.py (hunk 1 above)
Comment (GDay) [reply django_q/conf.py:50]: Anything that gets pushed to the branch would show up in the PR too. Doesn't look like you pushed anything to the `multi-queue` branch after I reviewed it. I did update your branch, so you might have to pull the changes I made first and then push the changes you made to Github.

Comment (sinowood) [reply django_q/conf.py:50]: Aha, that's true, please fix it. Thanks!

**Comment by GDay** [comment]
This is now released under v1.5.1! :partying_face:

**Pull request merged**